File: internal/transfer/transfer.go
package transfer

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"

	"github.com/schollz/progressbar/v3"

	"github.com/roidaradal/dali/internal/protocol"
)

// SendFile sends a file to the specified address
func SendFile(addr string, filePath string) error {
	// Open and get file info
	file, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	fileInfo, err := file.Stat()
	if err != nil {
		return fmt.Errorf("failed to get file info: %w", err)
	}

	fileSize := uint64(fileInfo.Size())
	filename := filepath.Base(filePath)

	// Connect to peer
	fmt.Printf("Connecting to %s...\n", addr)
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}
	defer conn.Close()

	// Send file offer
	offer := protocol.NewFileOfferMessage(filename, fileSize)
	_, err = conn.Write(offer.ToBytes())
	if err != nil {
		return fmt.Errorf("failed to send file offer: %w", err)
	}

	// Wait for response
	reader := bufio.NewReader(conn)
	responseLine, err := reader.ReadString('\n')
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	response, err := protocol.ParseTransferMessage([]byte(strings.TrimSpace(responseLine)))
	if err != nil {
		return fmt.Errorf("invalid response: %w", err)
	}

	switch response.Type {
	case "Accept":
		fmt.Printf("Peer accepted. Sending '%s'...\n", filename)
	case "Reject":
		fmt.Println("Peer rejected the file transfer.")
		return nil
	default:
		fmt.Println("Invalid response from peer.")
		return nil
	}

	// Send file data with progress bar
	bar := progressbar.NewOptions64(
		int64(fileSize),
		progressbar.OptionSetDescription("Sending"),
		progressbar.OptionSetWidth(40),
		progressbar.OptionShowBytes(true),
		progressbar.OptionShowCount(),
		progressbar.OptionSetTheme(progressbar.Theme{
			Saucer:        "█",
			SaucerHead:    "▓",
			SaucerPadding: "░",
			BarStart:      "[",
			BarEnd:        "]",
		}),
	)

	buf := make([]byte, protocol.ChunkSize)
	for {
		n, err := file.Read(buf)
		if err == io.EOF {
			break
		}
		if err != nil {
			return fmt.Errorf("failed to read file: %w", err)
		}

		_, err = conn.Write(buf[:n])
		if err != nil {
			return fmt.Errorf("failed to send data: %w", err)
		}
		bar.Add(n)
	}

	fmt.Println("\n✓ File sent successfully!")
	return nil
}

// ReceiveFiles listens for incoming file transfers
func ReceiveFiles(port uint16, outputDir string) error {
	listener, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", port))
	if err != nil {
		return fmt.Errorf("failed to listen: %w", err)
	}
	defer listener.Close()

	fmt.Printf("Listening for incoming files on port %d...\n", port)

	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}

		fmt.Printf("\nIncoming connection from %s\n", conn.RemoteAddr())

		go func(c net.Conn) {
			defer c.Close()
			if err := handleIncomingTransfer(c, outputDir); err != nil {
				fmt.Printf("Transfer error: %v\n", err)
			}
		}(conn)
	}
}

func handleIncomingTransfer(conn net.Conn, outputDir string) error {
	reader := bufio.NewReader(conn)

	// Read file offer
	offerLine, err := reader.ReadString('\n')
	if err != nil {
		return fmt.Errorf("failed to read offer: %w", err)
	}

	offer, err := protocol.ParseTransferMessage([]byte(strings.TrimSpace(offerLine)))
	if err != nil {
		return fmt.Errorf("invalid offer: %w", err)
	}

	if offer.Type != "FileOffer" {
		return fmt.Errorf("expected file offer, got %s", offer.Type)
	}

	filename := offer.Filename
	fileSize := offer.Size

	fmt.Printf("Receiving '%s' (%d bytes)\n", filename, fileSize)

	// Auto-accept for now
	accept := protocol.NewAcceptMessage()
	_, err = conn.Write(accept.ToBytes())
	if err != nil {
		return fmt.Errorf("failed to send accept: %w", err)
	}

	// Receive file data
	outputPath := filepath.Join(outputDir, filename)
	file, err := os.Create(outputPath)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	bar := progressbar.NewOptions64(
		int64(fileSize),
		progressbar.OptionSetDescription("Receiving"),
		progressbar.OptionSetWidth(40),
		progressbar.OptionShowBytes(true),
		progressbar.OptionShowCount(),
		progressbar.OptionSetTheme(progressbar.Theme{
			Saucer:        "█",
			SaucerHead:    "▓",
			SaucerPadding: "░",
			BarStart:      "[",
			BarEnd:        "]",
		}),
	)

	buf := make([]byte, protocol.ChunkSize)
	var received uint64

	for received < fileSize {
		toRead := protocol.ChunkSize
		if remaining := fileSize - received; remaining < uint64(toRead) {
			toRead = int(remaining)
		}

		n, err := reader.Read(buf[:toRead])
		if err != nil {
			if err == io.EOF {
				break
			}
			return fmt.Errorf("failed to read data: %w", err)
		}

		_, err = file.Write(buf[:n])
		if err != nil {
			return fmt.Errorf("failed to write file: %w", err)
		}

		received += uint64(n)
		bar.Add(n)
	}

	fmt.Printf("\n✓ Saved to '%s'\n", outputPath)
	return nil
}

File: src/transfer.rs
use crate::protocol::TransferMessage;
use indicatif::{ProgressBar, ProgressStyle};
use std::path::Path;
use tokio::fs::File;
use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
use tokio::net::{TcpListener, TcpStream};

const CHUNK_SIZE: usize = 64 * 1024; // 64KB chunks

/// Send a file to a peer
pub async fn send_file(
    addr: &str,
    file_path: &Path,
) -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(file_path).await?;
    let metadata = file.metadata().await?;
    let file_size = metadata.len();
    let filename = file_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown")
        .to_string();

    println!("Connecting to {}...", addr);
    let mut stream = TcpStream::connect(addr).await?;

    // Send file offer
    let offer = TransferMessage::FileOffer {
        filename: filename.clone(),
        size: file_size,
    };
    stream.write_all(&offer.to_bytes()).await?;

    // Wait for response
    let mut reader = BufReader::new(&mut stream);
    let mut response_line = String::new();
    reader.read_line(&mut response_line).await?;
    
    let response = TransferMessage::from_bytes(response_line.trim().as_bytes());

    match response {
        Some(TransferMessage::Accept) => {
            println!("Peer accepted. Sending '{}'...", filename);
        }
        Some(TransferMessage::Reject) => {
            println!("Peer rejected the file transfer.");
            return Ok(());
        }
        _ => {
            println!("Invalid response from peer.");
            return Ok(());
        }
    }

    // Send file data
    let pb = ProgressBar::new(file_size);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{spinner:.green} [{bar:40.cyan/blue}] {bytes}/{total_bytes} ({eta})")?
            .progress_chars("█▓░"),
    );

    let mut file = File::open(file_path).await?;
    let mut buf = vec![0u8; CHUNK_SIZE];
    let stream = reader.into_inner();

    loop {
        let n = file.read(&mut buf).await?;
        if n == 0 {
            break;
        }
        stream.write_all(&buf[..n]).await?;
        pb.inc(n as u64);
    }

    pb.finish_with_message("Transfer complete!");
    println!("✓ File sent successfully!");

    Ok(())
}

/// Listen for incoming file transfers
pub async fn receive_files(
    port: u16,
    output_dir: &Path,
) -> Result<(), Box<dyn std::error::Error>> {
    let listener = TcpListener::bind(format!("0.0.0.0:{}", port)).await?;
    println!("Listening for incoming files on port {}...", port);

    loop {
        let (mut stream, addr) = listener.accept().await?;
        println!("\nIncoming connection from {}", addr);

        let output_dir = output_dir.to_path_buf();

        tokio::spawn(async move {
            if let Err(e) = handle_incoming_transfer(&mut stream, &output_dir).await {
                eprintln!("Transfer error: {}", e);
            }
        });
    }
}

async fn handle_incoming_transfer(
    stream: &mut TcpStream,
    output_dir: &Path,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut reader = BufReader::new(stream);
    let mut offer_line = String::new();
    reader.read_line(&mut offer_line).await?;

    let offer = TransferMessage::from_bytes(offer_line.trim().as_bytes());

    let (filename, file_size) = match offer {
        Some(TransferMessage::FileOffer { filename, size }) => (filename, size),
        _ => {
            return Err("Invalid file offer".into());
        }
    };

    println!("Receiving '{}' ({} bytes)", filename, file_size);

    // Auto-accept for now
    let stream = reader.into_inner();
    let accept = TransferMessage::Accept;
    stream.write_all(&accept.to_bytes()).await?;

    // Receive file data
    let output_path = output_dir.join(&filename);
    let mut file = File::create(&output_path).await?;

    let pb = ProgressBar::new(file_size);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{spinner:.green} [{bar:40.cyan/blue}] {bytes}/{total_bytes} ({eta})")?
            .progress_chars("█▓░"),
    );

    let mut buf = vec![0u8; CHUNK_SIZE];
    let mut received: u64 = 0;

    while received < file_size {
        let to_read = std::cmp::min(CHUNK_SIZE as u64, file_size - received) as usize;
        let n = stream.read(&mut buf[..to_read]).await?;
        if n == 0 {
            break;
        }
        file.write_all(&buf[..n]).await?;
        received += n as u64;
        pb.inc(n as u64);
    }

    pb.finish_with_message("Complete!");
    println!("✓ Saved to '{}'", output_path.display());

    Ok(())
}
